[FLINK-2488][FLINK-2496] Expose Task Manager configuration and Task attempt number to Runtime context

sachingoel0101 opened the pull request:
This PR fixes these issues:
1. [FLINK-2496]Expose Task Manager configuration to Runtime Context 
2. [FLINK-2488]Expose attempt number of task to Runtime Context
3. [FLINK-2524]Add getTaskNameWithSubtasks to Runtime Context

Depends on #970 for simplistic constructors for `RuntimeContext`.

sachingoel0101 commented:
Since #970 is in, this can be merged now.

StephanEwen commented on flink-core/src/main/java/org/apache/flink/api/common/functions/util/AbstractRuntimeUDFContext.java:
```diff
@@ -93,6 +93,15 @@ public int getIndexOfThisSubtask() {
 	}
 
 	@Override
+	public String getTaskNameWithSubtasks() {
```
Since this method may be called rather often, I would create this string once and return it then.

sachingoel0101 replied:
Will fix this.

StephanEwen commented on flink-runtime/src/main/java/org/apache/flink/runtime/deployment/TaskDeploymentDescriptor.java:
```diff
@@ -120,7 +125,7 @@ public TaskDeploymentDescriptor(
 			List<BlobKey> requiredJarFiles, int targetSlotNumber) {
 
-		this(jobID, vertexID, executionId, taskName, indexInSubtaskGroup, numberOfSubtasks,
+		this(jobID, vertexID, executionId, taskName, indexInSubtaskGroup, numberOfSubtasks, 0,
```
I think such constructors with default values are dangerous, as people tend to call the wrong constructors.
Someone who creates the deployment descriptor should think about providing the attempt number.

sachingoel0101 replied:
This constructor was only being used in test classes, and the attempt number would've been kept 1 for them all. 
But yes. You're right. Leads to more consistency. Will fix this too.

StephanEwen commented:
Looks mostly good.

The `TaskRuntimeInfo` is now a bit strange. It contains some runtime info for tasks (attempt number) but not all (subtasks, etc). The attempt number is in the `TaskDeploymentDescriptor` anyways, so why copy it in addition to the `TaskRuntimeInfo`? Before, the `TaskManagerInfo` was clearly the context info of the TaskManager that was the same for all tasks and there was no duplicate information.

Looks like the motivation was to minimize the number of objects passed to the `RuntimeContext`. In that case, why not create a `RuntimeInfo` for the task (keep the `TaskManagerInfo`), put all the task-specific information in there, pass it to the `RuntimeEnvironment` and `RuntimeContext` and let them return all info like `getTaskName` and `getIndexOfThisSubtask` from there?

sachingoel0101 commented:
Yes. Minimizing the arguments being passed to `RuntimeContext` was the motivation. I thought about putting every task specific field into the TaskInfo object but since it isn't a real problem, hesitated. 
Should I do that then? I certainly like the idea. The constructors for `RuntimeContext` are somewhat messy, and doing this would make any future changes to contexts a lot easier.

StephanEwen commented:
Yes, let's have a `TaskRuntimeInfo` and a `TaskManagerContext`. Both are available in the `RuntimeEnvironment` and passed to the `RuntimeContext`.

The `TaskRuntimeInfo` has all task-specific information, the `TaskManagerContext` all cross-task constant parts. It could also hold the I/O manager, memory manager, ...

sachingoel0101 commented:
I have implemented a `TaskRuntimeInfo` which holds descriptive information about the task, such as name, index, parallelism, etc. Also renamed the `TaskManagerRuntimeInfo` to `TaskManagerContext` and included the `IOManager`, `MemoryManager` and `NetworkEnvironment`.

The `RuntimeContext`s now use these two fields in their constructor.
Further, for minimizing number of changes while modifying the `RuntimeContext`s in future, I have also modified the test classes to use a *create runtime* utility function.
Waiting for travis to give the green light for now.

sachingoel0101 commented:
Rebased to the latest master. Can be merged now.

StephanEwen commented on flink-core/src/main/java/org/apache/flink/api/common/TaskRuntimeInfo.java:
```diff
@@ -0,0 +1,103 @@
+	private final int attemptNumber;
+
+	/** Task Name along with information on index of task in group */
+	private transient final String taskNameWithSubTaskIndex;
```
If this is transient, it is null after deserialization. Either make it non-transient, or re-compute it after deserialization.

sachingoel0101 commented:
Fixed.

StephanEwen commented:
Sorry to be picky again, but it would be good to harmonize the code style a bit - make it more consistent with the remaining code.

What struck me in particular the omission of space before the curly braces at the start of methods and other code blocks. All other parts of the code have that.

There are ongoing discussions about making the code style stricter (also with respect to white spaces), and this is much easier if new code follows this standard already. Otherwise, there'd going to be a lot of style adjustments necessary when introducing the check.

sachingoel0101 commented:
Sure thing. I'll fix it. IntelliJ re-formats used to take care of such things, but I've stopped using it recently as it was messing up scala code.

sachingoel0101 commented:
@StephanEwen , I have fixed the styling issues as much as I could find out. 
Thanks for your patience. :)

sachingoel0101 commented:
Rebased to the current master. This should be mergeable now. Travis fails on unrelated kafka and flink-fs-tests. Just re-triggered another build.
@StephanEwen

sachingoel0101 commented:
Travis Passes successfully. https://travis-ci.org/apache/flink/builds/78368635

sachingoel0101 commented:
Will there be any further review of this?

StephanEwen commented:
Yeah, I'd like to add this once I find the time. Bit overloaded right now, please bear with us for a bit...

sachingoel0101 commented:
Sure. No worries. :)

sachingoel0101 commented:
Is it possible to get this in soon? I need access to the task manager configuration for something I'm working on. @StephanEwen

sachingoel0101 commented:
Unrelated failure on travis. Filed a jira [2711]

sachingoel0101 commented:
Closing this for now. Needs reworking.

StephanEwen commented:
Sorry for dropping this. The feature is still interesting. I have the feeling it can probably be implemented a bit simpler...

sachingoel0101 commented:
Yes. That's why I closed it. :')
I'm thinking of having a construct named `TaskInfo`, which contains information like name, index, parallel tasks, attempt number, etc. which will be passed all the way down from the `TDD` to the `RuntimeContext`. Let me know if that's a good idea.

StephanEwen commented:
Yes, a TaskInfo is a good idea.

It can be handed also to the `RuntimeEnvironment` (from there to the `RuntimeContext`), and is created once in the Task constructor from the TDD (that we we spare the JobManager from changes).

sachingoel0101 commented:
Agreed. 
I also would like to cleanup the several getter functions for these fields. We can just add a `getTaskInfo` in the `TDD`, `Task` and `Environment`. The fields can then be accessed from this object.
Since this isn't the user-facing API, it should be fine IMO. `RuntimeContext` will still provide separate access to every field.
